api/src/task/notifications/base_notification.py

```diff
@@ -0,0 +1,89 @@
+import logging
+from abc import abstractmethod
+from typing import Any
+from uuid import UUID
+
+import botocore.client
+from sqlalchemy import select
+
+from src.adapters.aws.pinpoint_adapter import send_pinpoint_email_raw
+from src.db.models.user_models import User, UserNotificationLog
+from src.task.notifications.constants import EmailData
+from src.task.task import Task
+
+logger = logging.getLogger(__name__)
+
+
+class BaseNotification(Task):
+
+    @abstractmethod
+    def collect_notifications(
+        self,
+    ) -> dict[UUID, list[Any]]:
+        """Collect notifications for users"""
+        pass
+
+    @abstractmethod
+    def prepare_notification(self, saved_data: dict[UUID, list[Any]]) -> EmailData:
+        """Prepare notification content (email data)"""
+        pass
+
+    def notification_data(self) -> EmailData | None:
+        """Fetch collected notifications and prepare email data."""
+        collected_data: dict[UUID, list[Any]] = self.collect_notifications()
+        if collected_data:
+            return self.prepare_notification(collected_data)
+        return None
+
+    def _get_user_email(self, user_id: UUID) -> str | None:
+
+        # Fetch user details
+        user = self.db_session.execute(
+            select(User).where(User.user_id == user_id)
+        ).scalar_one_or_none()
+
+        if not user or not user.email:
+            logger.warning(
+                "No email found for user", extra={"user_id": user.user_id if user else None}
+            )
+            return None
+        return user.email
```

Comment:
Since all notifications are driven by tables like `UserSavedOpportunity` or `UserSavedSearch` that are connected to the user, I wonder if there is a clean way of grabbing the user from those records?

That way we can both:
* Fetch the user object when fetching the saved record (better performance)
* Not need this special logic